src/painter.rs

```diff
@@ -0,0 +1,193 @@
+use crate::{
+    prompt::{PromptEditMode, PromptHistorySearch},
+    Prompt,
+};
+use crossterm::{
+    cursor::{self, position, MoveTo, MoveToColumn, RestorePosition, SavePosition},
+    style::{Color, Print, ResetColor, SetForegroundColor},
+    terminal::{self, Clear, ClearType},
+    QueueableCommand, Result,
+};
+
+use std::io::{Stdout, Write};
+
+pub struct Painter {
+    // Stdout
+    stdout: Stdout,
+}
+
+impl Painter {
+    pub fn new(stdout: Stdout) -> Self {
+        Painter { stdout }
+    }
+
+    pub fn queue_move_to(&mut self, column: u16, row: u16) -> Result<()> {
+        self.stdout.queue(cursor::MoveTo(column, row))?;
+
+        Ok(())
+    }
+
+    /// Queue the complete prompt to display including status indicators (e.g. pwd, time)
+    ///
+    /// Used at the beginning of each [`Reedline::read_line()`] call.
+    pub fn queue_prompt(&mut self, prompt: &dyn Prompt, prompt_mode: PromptEditMode) -> Result<()> {
+        let (screen_width, _) = terminal::size()?;
```

Comment:
Can also inject this? Not sure if this optimization is worth extending the interface that we provide.

Reply:
Reading the terminal::size is pretty slow, as behind the scenes crossterm has to write out an ansi sequence, wait, and read back the answer. We want to remember the terminal size we saw before and reuse it